File: spark/src/main/scala/geotrellis/spark/io/index/rowmajor/RowMajorSpatialKeyIndex.scala
package geotrellis.spark.io.index.rowmajor

import geotrellis.spark._
import geotrellis.spark.io.index._

import spire.syntax.cfor._

/** Represents a row major ordering for SpatialKey */
class RowMajorSpatialKeyIndex(keyBounds: KeyBounds[SpatialKey]) extends KeyIndex[SpatialKey] {
  val minCol = keyBounds.minKey.col
  val minRow = keyBounds.minKey.row
  
  val layoutCols = keyBounds.maxKey.col - keyBounds.minKey.col + 1

  println(layoutCols)

  def toIndex(key: SpatialKey): Long = 
    toIndex(key.row, key.col)

  def toIndex(row: Int, col: Int): Long =
    (layoutCols * (row - minRow) + (col - minCol)).toLong

  def indexRanges(keyRange: (SpatialKey, SpatialKey)): Seq[(Long, Long)] = {
    val SpatialKey(colMin, rowMin) = keyRange._1
    val SpatialKey(colMax, rowMax) = keyRange._2

    val cols = colMax - colMin + 1
    val rows = rowMax - rowMin

    val result = Array.ofDim[(Long, Long)](rowMax - rowMin + 1)

    cfor(0)(_ <= rows, _ + 1) { i =>
      val row = rowMin + i
      val min = toIndex(colMin, row)
      result(i) = (min, min + cols)
    }
    result
  }
}

File: geotrellis-python/geotrellis/spark/io/index/rowmajor/RowMajorSpatialKeyIndex.py
from geotrellis.spark.SpatialKey import SpatialKey
from geotrellis.spark.io.index.KeyIndex import KeyIndex

class RowMajorSpatialKeyIndex(KeyIndex[SpatialKey]):
    def __init__(self, keyBounds):
        self._keyBounds = keyBounds
        self.minCol = keyBounds.minKey.col
        self.minRow = keyBounds.minKey.row
        self.layoutCols = keyBounds.maxKey.col - keyBounds.minKey.col + 1

    def toIndex(self, col, row = None):
        if isinstance(col, SpatialKey) and row is None:
            key = col
            col, row = key.col, key.row
        return self.layoutCols * (row - self.minRow) + (col - self.minCol)

    def indexRanges(keyRange):
        minKey = keyRange[0]
        maxKey = keyRange[1]

        colMin = minKey.col
        rowMin = minKey.row
        colMax = maxKey.col
        rowMax = maxKey.row

        cols = colMax - colMin + 1
        rows = rowMax - rowMin

        result = [None] * (rows + 1)

        for i in xrange(0, rows):
            row = rowMin + i
            _min = self.toIndex(colMin, row)
            result[i] = (_min, _min + cols - 1)

        return result
